File: sdk/dart/lib/src/proxmox.dart
import 'http_client.dart';

/// HiTechCloudProxmox: machine status, power, VMs, IPs, backups, snapshots, bandwidth
class ProxmoxResource {
  final HttpClient httpClient;
  ProxmoxResource(this.httpClient);

  /// GET /api/service/{id}/htcpve/status - Get machine status
  Future<Map<String, dynamic>> getStatus([Map<String, dynamic>? params]) async {
    return httpClient.get('/api/service/{id}/htcpve/status', params);
  }

  /// POST /api/service/{id}/htcpve/power - Power action (start/stop/shutdown/reboot/reset)
  Future<Map<String, dynamic>> powerAction([Map<String, dynamic>? params]) async {
    return httpClient.post('/api/service/{id}/htcpve/power', params);
  }

  /// GET /api/service/{id}/htcpve/vms - List machines
  Future<Map<String, dynamic>> listVms([Map<String, dynamic>? params]) async {
    return httpClient.get('/api/service/{id}/htcpve/vms', params);
  }

  /// GET /api/service/{id}/htcpve/ips - List addresses
  Future<Map<String, dynamic>> listIps([Map<String, dynamic>? params]) async {
    return httpClient.get('/api/service/{id}/htcpve/ips', params);
  }

  /// POST /api/service/{id}/htcpve/rdns - Set reverse DNS
  Future<Map<String, dynamic>> setRdns([Map<String, dynamic>? params]) async {
    return httpClient.post('/api/service/{id}/htcpve/rdns', params);
  }

  /// GET /api/service/{id}/htcpve/backups - List backups
  Future<Map<String, dynamic>> listBackups([Map<String, dynamic>? params]) async {
    return httpClient.get('/api/service/{id}/htcpve/backups', params);
  }

  /// POST /api/service/{id}/htcpve/backups - Create backup
  Future<Map<String, dynamic>> createBackup([Map<String, dynamic>? params]) async {
    return httpClient.post('/api/service/{id}/htcpve/backups', params);
  }

  /// GET /api/service/{id}/htcpve/snapshots - List snapshots
  Future<Map<String, dynamic>> listSnapshots([Map<String, dynamic>? params]) async {
    return httpClient.get('/api/service/{id}/htcpve/snapshots', params);
  }

  /// POST /api/service/{id}/htcpve/snapshots - Create snapshot
  Future<Map<String, dynamic>> createSnapshot([Map<String, dynamic>? params]) async {
    return httpClient.post('/api/service/{id}/htcpve/snapshots', params);
  }

  /// GET /api/service/{id}/htcpve/usage - Get bandwidth usage
  Future<Map<String, dynamic>> getBandwidthUsage([Map<String, dynamic>? params]) async {
    return httpClient.get('/api/service/{id}/htcpve/usage', params);
  }

}

File: sdk/rust/src/proxmox.rs
//! HiTechCloudProxmox: machine status, power, VMs, IPs, backups, snapshots, bandwidth

use std::sync::Arc;
use crate::http_client::HttpClient;

pub struct ProxmoxResource {
    http: Arc<HttpClient>,
}

impl ProxmoxResource {
    pub fn new(http: Arc<HttpClient>) -> Self { Self { http } }

    /// GET /api/service/{id}/htcpve/status - Get machine status
    pub async fn get_status(&self, params: Option<&serde_json::Value>) -> Result<serde_json::Value, crate::http_client::ApiError> {
        self.http.get("/api/service/{id}/htcpve/status", None).await
    }

    /// POST /api/service/{id}/htcpve/power - Power action (start/stop/shutdown/reboot/reset)
    pub async fn power_action(&self, params: Option<&serde_json::Value>) -> Result<serde_json::Value, crate::http_client::ApiError> {
        self.http.post("/api/service/{id}/htcpve/power", None).await
    }

    /// GET /api/service/{id}/htcpve/vms - List machines
    pub async fn list_vms(&self, params: Option<&serde_json::Value>) -> Result<serde_json::Value, crate::http_client::ApiError> {
        self.http.get("/api/service/{id}/htcpve/vms", None).await
    }

    /// GET /api/service/{id}/htcpve/ips - List addresses
    pub async fn list_ips(&self, params: Option<&serde_json::Value>) -> Result<serde_json::Value, crate::http_client::ApiError> {
        self.http.get("/api/service/{id}/htcpve/ips", None).await
    }

    /// POST /api/service/{id}/htcpve/rdns - Set reverse DNS
    pub async fn set_rdns(&self, params: Option<&serde_json::Value>) -> Result<serde_json::Value, crate::http_client::ApiError> {
        self.http.post("/api/service/{id}/htcpve/rdns", None).await
    }

    /// GET /api/service/{id}/htcpve/backups - List backups
    pub async fn list_backups(&self, params: Option<&serde_json::Value>) -> Result<serde_json::Value, crate::http_client::ApiError> {
        self.http.get("/api/service/{id}/htcpve/backups", None).await
    }

    /// POST /api/service/{id}/htcpve/backups - Create backup
    pub async fn create_backup(&self, params: Option<&serde_json::Value>) -> Result<serde_json::Value, crate::http_client::ApiError> {
        self.http.post("/api/service/{id}/htcpve/backups", None).await
    }

    /// GET /api/service/{id}/htcpve/snapshots - List snapshots
    pub async fn list_snapshots(&self, params: Option<&serde_json::Value>) -> Result<serde_json::Value, crate::http_client::ApiError> {
        self.http.get("/api/service/{id}/htcpve/snapshots", None).await
    }

    /// POST /api/service/{id}/htcpve/snapshots - Create snapshot
    pub async fn create_snapshot(&self, params: Option<&serde_json::Value>) -> Result<serde_json::Value, crate::http_client::ApiError> {
        self.http.post("/api/service/{id}/htcpve/snapshots", None).await
    }

    /// GET /api/service/{id}/htcpve/usage - Get bandwidth usage
    pub async fn get_bandwidth_usage(&self, params: Option<&serde_json::Value>) -> Result<serde_json::Value, crate::http_client::ApiError> {
        self.http.get("/api/service/{id}/htcpve/usage", None).await
    }

}
